### PR TITLE
Fix Qwen 3.5 35B RL gibberish output issue and show end-to-end RL convergence

### PR DESCRIPTION
**Pull request by AntonyMei**
# Description

This PR fixes the Qwen 3.5 RL gibberish output issue and shows end-to-end RL training convergence with Qwen 3.5 35B + v5p-64 on GSM8K. This is a clean up of the fixes in yixuanm-dev-35b-new branch.

FIXES: b/521604343
FIXES: b/542769108

# Tests

On yixuanm-dev-35b-new, Qwen 3.5 35B + v5p-64 on GSM8K shows end-to-end convergence:
gs://yixuanm-maxtext-logs/ConvergenceTesting/ym-qw35-35/tensorboard

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/post_train/rl_gsm8k_qwen35_35b_v5p64.yml [hunk 1]
```diff
@@ -0,0 +1,159 @@
+# Copyright 2026 Google LLC
```
Comment (A9isha) [inline src/maxtext/configs/post_train/rl_gsm8k_qwen35_35b_v5p64.yml]: we can skip this file.

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py [hunk 2]
```diff
@@ -414,31 +414,12 @@ def patched_get_kv_cache_spec(self):
...
-        torch_dtype = torch.float16
-      mamba_dtypes = (torch_dtype, torch_dtype)
-
-      # Calculate unpadded mamba page size
-      dtype_size = 4 if torch_dtype == torch.float32 else 2
-      unpadded_mamba_page_size = sum(int(np.prod(shape)) * dtype_size for shape in mamba_shapes)
+      # Qwen GDN keeps its short convolution history in BF16, but recurrence is
+      # accumulated and persisted in FP32. Declaring both caches as the model
+      # dtype silently quantizes the recurrent state after every generated token.
+      mamba_shapes, mamba_dtypes, unpadded_mamba_page_size = build_qwen_gdn_cache_layout(
```
Comment (A9isha) [inline src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py]: but this file is general across all models. You probably want to make this `build_qwen_gdn_cache_layout` conditional on the model

File: src/maxtext/integration/vllm/hybrid_cache_utils.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Hybrid cache-layout helpers for MaxText's vLLM adapter."""
+
+import math
+from typing import Any
+
+
+def build_qwen_gdn_cache_layout(cfg: Any, torch_module: Any):
```
Comment (A9isha) [inline src/maxtext/integration/vllm/hybrid_cache_utils.py:38]: is this qwen specific? could other models also need such cache utilities?

File: src/maxtext/layers/embeddings.py [hunk 4]
```diff
@@ -1792,28 +1793,42 @@ def __init__(
     Args:
       min_timescale: Start of the geometric index (typically 1).
       max_timescale: End of the geometric index (rope_theta, e.g., 1000000).
-      embedding_dims: Dimension of the embedding (head_dim).
+      embedding_dims: Dimension of the attention head.
```
Comment (A9isha) [inline src/maxtext/layers/embeddings.py:1824]: now that https://github.com/AI-Hypercomputer/maxtext/pull/4764 is merged, please rebase to remove those change from this PR

File: src/maxtext/utils/sharding.py [hunk 5]
```diff
@@ -910,6 +910,52 @@ def remove_mesh_axes_from_partition_spec(pspec, axes_to_remove, dims=None):
   return jax.sharding.PartitionSpec(*new_spec)
 
 
+def remove_incompatible_mesh_axes_from_partition_spec(pspec, shape, mesh, dims=None):
```
Comment (NuojCheng) [inline src/maxtext/utils/sharding.py]: I am worried this function gonna silently fallback sharding configs instead of shouting out and throwing errors. I suggested adding a flag `allow_remove_axes=False` and this fallback happens only when the flag is true.

File: src/maxtext/layers/moe.py [hunk 6]
```diff
@@ -42,8 +42,15 @@
...
-from maxtext.utils.sharding import logical_to_mesh_axes, remove_expert_from_partition_spec, get_logical_axis_rules
+from maxtext.utils.sharding import (
+    create_sharding,
+    get_logical_axis_rules,
+    logical_to_mesh_axes,
+    maybe_shard_with_logical,
+    maybe_shard_with_pspec,
+    remove_expert_from_partition_spec,
+    remove_incompatible_mesh_axes_from_partition_spec,
+)
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py:54]: nice clean up, thank you

File: src/maxtext/layers/moe.py [hunk 7]
```diff
@@ -1647,6 +1654,48 @@ def get_routed_moe_shardings(is_batch_sharded_by_expert, has_input_ids):
...
+            input_ids.shape,
+            self.mesh,
+            dims=(0,),
+        )
+      output_pspec = remove_incompatible_mesh_axes_from_partition_spec(
+          output_pspec,
+          inputs.shape,
+          self.mesh,
+          dims=(0,),
+      )
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py]: could you move all logics from 1664-1698 to a separate small function, maybe named `maybe_remove_incompatible_mesh_axes_from_partition_spec`, so that readers don't need to read this logic unless necessary

**Review by NuojCheng — APPROVED** [review]
Sharding and MoE components LGTM if my comments are resolved. Thank you for the code cleanup Yixuan!

File: src/maxtext/integration/vllm/maxtext_vllm_rollout.py [hunk 8]
```diff
@@ -64,6 +117,155 @@ def _find_scanned_layer_idx(key_tuple, container_names=("layers", "scanned_block
...
+  for i in range(len(key_tuple) - 1):
+    if key_tuple[i] != "layers" or not isinstance(key_tuple[i + 1], str):
+      continue
+    match = re.fullmatch(r"layer_(\d+)", key_tuple[i + 1])
+    if match:
+      return i, int(match.group(1))
+  return -1, -1
+
+
+def unroll_qwen_scanned_weights(weights, scan_axis: int = 1, pattern_length: Optional[int] = None):
```
Comment (khatwanimohit) [inline src/maxtext/integration/vllm/maxtext_vllm_rollout.py:131]: can you use https://github.com/AI-Hypercomputer/maxtext/blob/1555e1feffc4ea6e49dcfa6bb76ae6784cfdff14/src/maxtext/utils/max_utils.py#L1064 for unrolling params ? unless there is something unique about qwen3 scanned weights which I don't know about

File: src/maxtext/utils/sharding.py [hunk 9]
```diff
@@ -910,6 +910,52 @@ def remove_mesh_axes_from_partition_spec(pspec, axes_to_remove, dims=None):
...
+    dims: Dim indices to check; `None` (the default) checks every dim.
+
+  Returns:
+    A PartitionSpec whose checked dimensions are evenly shardable.
+  """
+  if len(pspec) > len(shape):
+    raise ValueError(f"PartitionSpec rank {len(pspec)} exceeds tensor rank {len(shape)}")
+
+  dims_to_check = None if dims is None else set(dims)
+  compatible_pspec = pspec
```
Comment (NuojCheng) [inline src/maxtext/utils/sharding.py:1031]: what if dims < 0?

File: tests/post_training/unit/extract_answer_test.py [hunk 10]
```diff
@@ -102,6 +103,67 @@ def test_legacy_last_answer_wins(self):
...
+    config = SimpleNamespace(
+        reasoning_start_token="<think>",
+        reasoning_end_token="</think>",
+        reasoning_start_token_in_prompt=True,
+        solution_start_token="<answer>",
+        solution_end_token="</answer>",
+    )
+    response = "We calculate 6 * 7.</think>\nThe final answer is 42. do do do"
+
+    self.assertEqual(utils_rl.extract_answer(response, config), "42")
```
Comment (A9isha) [inline tests/post_training/unit/extract_answer_test.py]: I know the `native final answer` helps when the model did utter the answer and it is frustrating to not give it the reward. but it was not within answer tags, but if we allow it to put the answer this way it will not learn the format well, I know in the easy gsm8k dataset it is learning the format finally after ~150 training steps

**Review by YixuanWang-99 — APPROVED** [review]

File: src/maxtext/configs/post_train/rl_gsm8k_qwen35_35b_v5p64.yml (hunk 1 above)
Comment (AntonyMei) [reply src/maxtext/configs/post_train/rl_gsm8k_qwen35_35b_v5p64.yml]: Sure, I will remove the file from the current PR and submit it in another PR.

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py (hunk 2 above)
Comment (AntonyMei) [reply src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py]: Currently, this function call is under "if decoder_block_str in ("qwen3_next", "qwen3_5")", so it is already model specific.

File: src/maxtext/integration/vllm/hybrid_cache_utils.py (hunk 3 above)
Comment (AntonyMei) [reply src/maxtext/integration/vllm/hybrid_cache_utils.py:38]: This function is Qwen specific, but I will change the file name so that later on we can put helper functions for other models here as well.

File: src/maxtext/utils/sharding.py (hunk 5 above)
Comment (AntonyMei) [reply src/maxtext/utils/sharding.py]: Sure, I will add a flag and default it to false.

File: src/maxtext/layers/moe.py (hunk 7 above)
Comment (AntonyMei) [reply src/maxtext/layers/moe.py]: Sure, will do.

File: src/maxtext/utils/sharding.py (hunk 9 above)
Comment (AntonyMei) [reply src/maxtext/utils/sharding.py:1031]: Will add support for negative indexing.

File: src/maxtext/layers/embeddings.py (hunk 4 above)
Comment (AntonyMei) [reply src/maxtext/layers/embeddings.py:1824]: Sure.

File: tests/post_training/unit/extract_answer_test.py (hunk 10 above)
Comment (AntonyMei) [reply tests/post_training/unit/extract_answer_test.py]: Got it, I will remove the new numeric-prose fallback.

File: src/maxtext/integration/vllm/maxtext_vllm_rollout.py (hunk 8 above)
Comment (AntonyMei) [reply src/maxtext/integration/vllm/maxtext_vllm_rollout.py:131]: `max_utils.unscan_train_state_params` is not applicable here because it expects a Linen params/sharding pair with homogeneous layer groups; this path receives an NNX state and must interleave heterogeneous layer slots. I have added this note in the docstring.

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py (hunk 2 above)
Comment (A9isha) [reply src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py]: True. Thanks!

**Review by A9isha — APPROVED** [review]
Thanks a lot Yixuan!

**Review by A9isha — APPROVED** [review]
